### PR TITLE
test: add async generator test for fetch

### PR DESCRIPTION
**Pull request by zmzlois**
add async generator test in Bun 
~~part 2 double retention in startRequestStream~~

~~when starting the streaming of a request body `startRequestStream`, it kept the original `request_body.ReadableStream` Strong in the `HTTPRequestBody` union while also creating and keeping a `ResumableFetchSink` that references the same stream. In `clearData`, the union’s ReadableStream is not detached once streaming is started (“do not detach if .ReadableStream and not waiting to start”). very likely that its retaining an extra Strong reference unnecessarily --> leaking (the stream hangs around even when the sink is torn down).~~

### What does this PR do?
- transfer ownership of the Strong from `request_body` into the sink when starting, so the union no longer holds it.
- detach `request_body` in all cases where sink is `null` in `clearData()` , and let `clearSink()` handle sink teardown (and thus releasing the only remaining Strong).

### How did you verify your code works?
once this the PR is open will use this in our canary to test in production 

TODO: 
- [x] part 1 https://github.com/oven-sh/bun/pull/24247 superseded by https://github.com/oven-sh/bun/pull/24251 
- [x] part 2 fixes for double retention in `startRequestStream`
- [ ] part 3: Response finalizer logic/comments mismatch and body ignoring heuristics
- [ ] part 4: S3 streaming + AbortSignal not wired for cancellation 
- [ ] part 5: `has_schedule_callback` atomic usage and scheduling robustness
- [ ] part 6: Body buffering allocator handoff 
- [ ] part 7: edge cases for Request-body “ReadableStream not started yet”

### CODE REVIEW

File: src/bun.js/webcore/fetch.zig [hunk 1]
```diff
@@ -341,10 +346,18 @@ pub const FetchTasklet = struct {
     pub fn startRequestStream(this: *FetchTasklet) void {
         this.is_waiting_request_stream_start = false;
         bun.assert(this.request_body == .ReadableStream);
-        if (this.request_body.ReadableStream.get(this.global_this)) |stream| {
+
+        // Making sure the stream doesn't retain twice -transfer ownership out of the union to avoid double-Strong retention
+        var stream_ref = &this.request_body.ReadableStream;
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/fetch.zig]: you probably meant to copy it by value instead of keeping a `&` pointer. Assigning to `this.request_body` reassigns the pointer, so on the next line this `stream_ref` becomes invalid memory.
```suggestion
        var stream_ref = this.request_body.ReadableStream;
```

File: src/bun.js/webcore/fetch.zig [hunk 2]
```diff
@@ -341,10 +346,18 @@ pub const FetchTasklet = struct {
...
+        this.request_body = HTTPRequestBody.Empty;
+
+        if (stream_ref.get(this.global_this)) |stream| {
             if (this.signal) |signal| {
                 if (signal.aborted()) {
+                    // match what was happenign earlier but owning the Strong in current scope
                     stream.abort(this.global_this);
+                    // drop transferred strong
+                    stream_ref.deinit();
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/fetch.zig]: Since each branch in this function is calling `deinit`, we can instead add a `defer stream_ref.deinit()` at line 351 where it's declared and then we can remove the duplicate code.

File: src/bun.js/webcore/fetch.zig [hunk 3]
```diff
@@ -354,7 +367,11 @@ pub const FetchTasklet = struct {
             // +1 because the task refs the sink
             const sink = ResumableSink.initExactRefs(globalThis, stream, this, 2);
             this.sink = sink;
+            // sink now owns the stream; drop the transferred Strong
+            stream_ref.deinit();
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/fetch.zig]:
```suggestion

```

File: src/bun.js/webcore/fetch.zig [hunk 4]
```diff
@@ -354,7 +367,11 @@ pub const FetchTasklet = struct {
...
         }
+        // in any other cases we will drop the transferred Strong anyways
+        stream_ref.deinit();
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/fetch.zig]:
```suggestion

```

File: test/js/web/fetch/fetch-readablestream-request-body.test.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      port: 0,
+      async fetch() {
+        await Bun.sleep(100);
+        return new Response("ok");
+      },
+    });
+
+    const request = new Request(server.url, {
+      method: "POST",
+      body: "test data that might become a stream",
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-readablestream-request-body.test.ts]: This will never reach the code path because `body` does not become a `ReadableStream` and we use a separate optimization that detects the data is buffered. So you could probably delete this test.

File: test/js/web/fetch/fetch-readablestream-request-body.test.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      },
+    });
+
+    const originalRequest = new Request(server.url, {
+      method: "POST",
+      body: "test data",
+    });
+
+    // clone creates a new request that might use ReadableStream internally
+    const clonedRequest = originalRequest.clone();
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-readablestream-request-body.test.ts]: This will also not exercise the intended code path because we do a clever thing where we increment the reference count of the underlying `WTF::StringImpl`.

File: src/bun.js/webcore/fetch.zig [hunk 7]
```diff
@@ -298,6 +298,11 @@ pub const FetchTasklet = struct {
             this.request_body.detach();
         }
 
+        // if sink exists, it owns any request ReadableStream - but if not then detach the request body
+        if (this.sink == null) {
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/fetch.zig]: Instead, can we move this up to the line above on 297 (while keeping the comment)?

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Some cleanup to do but it looks plausible. 

To write a test that exercises this code path, I suggest a test that waits for `pull` to be called on the `ReadableStream` and then immediately .abort() the signal inside the pull callback?

File: test/js/web/fetch/fetch-readablestream-request-body.test.ts (hunk 6 above)
Comment (zmzlois) [reply test/js/web/fetch/fetch-readablestream-request-body.test.ts]: that `WTF` looks interesting...

File: test/js/web/fetch/fetch-readablestream-request-body.test.ts [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+        return new Response(text);
+      },
+    });
+
+    const chunk = new TextEncoder().encode("test data");
+
+    const originalRequest = new Request(server.url, {
+      method: "POST",
+      body: new ReadableStream({
+        async start(controller) {
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-readablestream-request-body.test.ts]: This test should use `async pull` with a delay of say 0ms because otherwise we may detect that the ReadableStream has already completed by the time the `fetch` callback is called.

File: test/js/web/fetch/fetch-readablestream-request-body.test.ts [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+      port: 0,
+      async fetch(req) {
+        const text = await req.text();
+        return new Response(text);
+      },
+    });
+
+    const makeStream = () =>
+      new ReadableStream({
+        async start(controller) {
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-readablestream-request-body.test.ts]: This test should also use `pull`

**Comment by zmzlois** [comment]
note for self: this is the test i need to pay attention to on next step https://github.com/oven-sh/bun/pull/24276/commits/687421bea8aa98c39d3c4b77686cb0fe2cc6f55d
